### DISCUSSION TITLE
Ask after execution: Hotkeys for buttons (was: Tab key navigation between form elements)

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#139] [Ideas - done & released]
Edit 2022-03-05: This original post talks about tab key navigation. See this comment for the up-to-date idea: https://github.com/Taitava/obsidian-shellcommands/discussions/139#discussioncomment-2083497

## Original post

I opened this new discussion to continue talking about an idea by @FelipeRearden.

@FelipeRearden https://github.com/Taitava/obsidian-shellcommands/discussions/138#discussioncomment-2076204:
> 
> 02 - Redirect: Support for Keyboard navigation using TAB to switch between outputs
> 
>   * if possible change the `border-color` of the selected output
> 
> 
> 03- Redirect: CRTL/CMD+ENTER to select the output that was choose by item 2

Being able to tab through buttons in the UI makes sense. The only thing I'm thinking is the scope of it: it's a larger thing related to more than just the output modal (#134). All other modals should also have the ability to switch focus between elements by using tab, and to trigger buttons by pressing Enter (maybe with a modififier key like you suggested). Well, now that I tested, I noticed that tab key works to switch between text fields and dropdown fields. On/off toggle fields, number range fields and buttons are not navigateable by pressing the tab key.

IMO the scope of this idea is Obsidian as a whole, so I'm thinking that we should make a feature request to them. If they implement it, I'd guess it would come to the output modal automatically, without any changes in this plugin. If they would not do it for some reason, then I could try to research it (with the goal to apply it for all modals in SC) sometime 🙂. It's a really good idea, and I definitely support it! 😎👍

**Comment by FelipeRearden** [comment]
> Being able to tab through buttons in the UI makes sense. The only thing I'm thinking is the scope of it: it's a larger thing related to more than just the output modal (#134). All other modals should also have the ability to switch focus between elements by using tab, and to trigger buttons by pressing Enter (maybe with a modififier key like you suggested).

Let me share why/how I had this idea of using keyboard hotkeys to Modal. 

The **RSS Reader** plugin and **From Template** plugin uses a modal window as well. In their cases it was possible to define specific hotkeys to trigger the commands from buttons and text field. The links are from my FR from these plugins.  
- https://github.com/joethei/obsidian-rss/issues/13
- https://github.com/mo-seph/obsidian-note-from-template/issues/9

Another good example is the beta plugin Search Core Assistant 
Take a look at the hotkey options for the Modal (it opens a window based on what you search with the Search core plugin)
- https://github.com/qawatake/obsidian-core-search-assistant-plugin#manual-preview

> Well, now that I tested, I noticed that tab key works to switch between text fields and dropdown fields. On/off toggle fields, number range fields and buttons are not navigateable by pressing the tab key.

Another idea !!!!: how about we use CMD/CRTL+1 , CMD/CRTL+2 , CMD/CRTL+3 , CMD/CRTL+4 and CMD/CRTL+5 to trigger each one of the Redirect commands instead of TAB?????? 

> IMO the scope of this idea is Obsidian as a whole, so I'm thinking that we should make a feature request to them. If they implement it, I'd guess it would come to the output modal automatically, without any changes in this plugin. If they would not do it for some reason, then I could try to research it (with the goal to apply it for all modals in SC) sometime 🙂. It's a really good idea, and I definitely support it! 😎👍 

Let'ssee what you think about what I wrote above :) 

But I agree with you, would be amazing if Obsidian implemented more native features for Modals 

Let keep talking :)

**Reply by Taitava** [reply]
> The **RSS Reader** plugin and **From Template** plugin uses a modal window as well. In their cases it was possible to define specific hotkeys to trigger the commands from buttons and text field. The links are from my FR from these plugins.
> 
>   * [[Feature Request] Hotkeys for Navigation inside the visualization window joethei/obsidian-rss#13](https://github.com/joethei/obsidian-rss/issues/13)
> 
>   * [[Feature Request] `Add`Button-> Keyboard Shortcut `CRTL/CMD+ENTER` to hit Add mo-seph/obsidian-note-from-template#9](https://github.com/mo-seph/obsidian-note-from-template/issues/9)

Thank you for these links! I looked at the source code of _Notes from template_ and was able to find out how to add this kinf of context related hotkeys in Obsidian API. I'll write it here just as a note to myself: `Modal.scope.register()`.


> 
> 
> Another good example is the beta plugin Search Core Assistant Take a look at the hotkey options for the Modal (it opens a window based on what you search with the Search core plugin)
> 
>   * https://github.com/qawatake/obsidian-core-search-assistant-plugin#manual-preview

Sorry, I won't try this now as it's not yet in the official list of community plugins.

> Another idea !!!!: how about we use CMD/CRTL+1 , CMD/CRTL+2 , CMD/CRTL+3 , CMD/CRTL+4 and CMD/CRTL+5 to trigger each one of the Redirect commands instead of TAB??????

Yea, I like this idea more than the _use tab to switch between buttons_. I mean, the latter is still a valid idea, and I hope it will be implemented in Obsidian (FR needs to be done, if a similar one is not yet present in the forums - could you try to search for one, please? 🙂 ). But as for something to be implemented in this plugin, and for this particular modal, this _CMD/CRTL+1 , CMD/CRTL+2 etc..._ idea is the one I would go forward with.

The only issue I see with the hotkey **numbers** is that they will change when new output channels get added. But then again, I don't see better replacements for them in letter keys.

**Reply by FelipeRearden** [reply]
> Yea, I like this idea more than the _use tab to switch between buttons_. I mean, the latter is still a valid idea, and I hope it will be implemented in Obsidian (FR needs to be done, if a similar one is not yet present in the forums - could you try to search for one, please? 🙂 ). But as for something to be implemented in this plugin, and for this particular modal, this _CMD/CRTL+1 , CMD/CRTL+2 etc..._ idea is the one I would go forward with.

Thank you @Taitava !!!!!

I'm glad that you liked!!!!!!

**Comment by Taitava** [comment]
## Will be released in `0.12.0`

This is how it looks like:

![kuva](https://user-images.githubusercontent.com/13002142/156883892-de64f618-a5b8-429b-a304-c8c0cbbbdb2a.png)

Any comments on the tooltip text? Is it good or does it need improving?

**Reply by FelipeRearden** [reply]
Hello @Taitava !!!!!

My humble opinion ...

```
Redirect: Normal Click OR CTRL+5
<br>
Redirect and Close Modal: CTRL Click + Normal Click OR CTRL+ALT+5
```

OR (like like this one more) 

```
Status Bar
<br>
Redirect: Normal Click OR CTRL+5
<br>
Redirect and Close Modal: CTRL Click + Normal Click OR CTRL+ALT+5
```

Would be great if we have one line for each helper 

Have a great day

**Reply by Taitava** [reply]
Sorry, I missed your message! 😦 Good suggestion, I'll change the tooltip!

**Reply by FelipeRearden** [reply]
> Sorry, I missed your message! 😦 Good suggestion, I'll change the tooltip!

No problem @Taitava !!!! Git messes up when we have to conversations in the same post :)

Git is not powerful enough for us 😎 🤠 

I'm glad you liked !!!!!

**Reply by Taitava** [reply]
> No problem @Taitava !!!! Git messes up when we have to conversations in the same post :)

Actually, it's _GitHub_ 😉 . [Git](https://git-scm.com) is a source code version control program, and GitHub is this website that can be used to host projects that use Git. So, Git is used locally on a computer (just like Obsidian), and GitHub offers additional services in the internet. For example, these discussions and issues are not part of Git itself, but they are something that GitHub has implemented. Then again, [commits](https://github.com/Taitava/obsidian-shellcommands/commits/main) and [branches](https://github.com/Taitava/obsidian-shellcommands/branches) are a core part of Git, and they can also be accessed in GitHub, too. There are also other services similar to GitHub, that utilizes Git in their service, such as [GitLab](https://about.gitlab.com/) and [Bitbucket](https://bitbucket.org/product/).

That being said, I do agree with you that the discussions feature in GitHub does not indicate very well if a discussion has new messages in two places on the same page. You are correct there. 👍

**Reply by FelipeRearden** [reply]
Thank you very much for the explanation @Taitava !!!!

I'm happy that you explain this to me :) as you know this is a new world for me and it's good to learn the right terms to make the communication easier!

Have a great day!!!

**Reply by Taitava** [reply]
I'm glad I can help! 🙂 It's easy to confuse Git and GitHub. 🙂

Have a nice day! 🌞

**Comment by Taitava** [comment]
# Small problem with numpad

@FelipeRearden `Ctrl` + `Shift` + a number on numpad does not work at least on Windows, because it starts to use numpad keys as arrow keys (8 becomes _up_, 4 becomes _left_, 6 becomes _right_, and 2 becomes _down_). I have Num Lock turned **on**, so I assumed numpad would press numbers instead of arrow keys even if `Ctrl` and `Shift` are pressed the same time, but I was wrong. If I press just `Ctrl` + a number on numpad, then the hotkey works.

We have two options:
a) Accept that when using `Ctrl`/`Cmd` + `Shift`, we need to use normal number keys, not numpad? That works correctly, but I just hate having to say to users who might ask about this, that it cannot be fixed.
b) Change `Shift` to `Alt`? So the combination would be `Ctrl`/`Cmd` + `Alt` + a number. But I don't like `Alt` so much.

What would you think?

I don't know how this works in macOS.

**Reply by FelipeRearden** [reply]
> We have two options:
> 
> a) Accept that when using `Ctrl`/`Cmd` + `Shift`, we need to use normal number keys, not numpad? That works correctly, but I just hate having to say to users who might ask about this, that it cannot be fixed.

100% agree. 

> b) Change `Shift` to `Alt`? So the combination would be `Ctrl`/`Cmd` + `Alt` + a number. But I don't like `Alt` so much.

Sorry for my Mistake about CRTL+SHIFT on numpad . I forgot this detail :)

> What would you think?

I think our least worst option is option b . 

> I don't know how this works in macOS.

Same behave of Win happens on MacOS :)


Have a great day!!!!!!!

**Reply by Taitava** [reply]
Thank you, I'll change it to use `Alt` instead of `Shift`. That's our _**Alt**ernative_ solution! 😉

**Comment by Taitava** [comment]
And just one more question that came to my mind now that I've spent some time implementing and testing this.

## Are number keys ok, or should letters be used instead?

- When new output channels get added, I might add them in between the current channels, which means that the numbering can change, e.g. number `5` is currently assigned to `Status bar`, but if I add a new output channel before `Status bar`, it's number will change from `5` to `6`. Users need to learn the changed hotkey.
- If there will ever be more than 10 output channels, some channels will not have a hotkey. This is a future problem, i.e. can be solved later, if needed.

Letters could be e.g.
- `Ctrl`/`Cmd` + (`Shift`) + `R` for `Current file: ca[R]et position`.
- `Ctrl`/`Cmd` + (`Shift`) + `T` for `Current file: [T]op`.
- `Ctrl`/`Cmd` + (`Shift`) + `B` for `Current file: [B]ottom`.
- `Ctrl`/`Cmd` + (`Shift`) + `O` for `[O]pen a file`.
- `Ctrl`/`Cmd` + (`Shift`) + `S` for `[S]tatus bar`.
- `Ctrl`/`Cmd` + (`Shift`) + `L` for `C[L]ipboard`. (This needs to be other than `C`, because `Ctrl` + `C` is needed for normal copying when text is selected in the textarea field).

This can also be decided/changed later.

**Reply by Taitava** [reply]
@FelipeRearden do you have an opinion? 🙂

**Reply by FelipeRearden** [reply]
> @FelipeRearden do you have an opinion? 🙂 

100% agree with you @Taitava !!!! 

I vote for letters too. This way we don't have to worry about this in the future :)

And we have to avoid CRTL+SHIFT+I that opens the Console :)

**Reply by FelipeRearden** [reply]
This is my opinion about the tooltip UI 

https://github.com/Taitava/obsidian-shellcommands/discussions/139#discussioncomment-2301211

:)

Have a great day!!!!!

**Reply by Taitava** [reply]
> > @FelipeRearden do you have an opinion? 🙂
> 
> 100% agree with you @Taitava !!!!
> 
> I vote for letters too. This way we don't have to worry about this in the future :)
> 
> And we have to avoid CRTL+SHIFT+I that opens the Console :)

Good, I'll change them to letters. And I'll avoid `Ctrl`/`Cmd` + `Shift` + `I` in future output channels too. I use those keys a lot to open up the console, so I guess I will remember to avoid assigning that one. 🙂👍 

Edit: Have a great day, too! 😎

**Comment by Taitava** [comment]
# Changed hotkeys to letters
![kuva](https://user-images.githubusercontent.com/13002142/162607645-846fa48c-bd76-44c2-96b8-2936d8179e2b.png)

When changing numbers to letters, I removed them from the button labels. I think letters would only make the buttons look messy. The letters can be seen from the tooltips anyway.
